File: CSharp/Sudoku.Console/Models/Table.cs
using System;
using System.Text;

namespace Sudoku.Models;

public class Table
{
    private Number?[,] _numbers = new Number?[9, 9];
    private int noOfSetNumbers = 0;

    public Table(Template template)
	{
        for(int column = 0; column < 9; column++)
        {
            for (int row = 0; row < 9; row++) 
            {
                if (template.Grid[column][row] != 0) 
                {
                    _numbers[row, column] = new(template.Grid[column][row], false);
                }
            }
        }
	}

    public bool IsFilled()
    {
        return noOfSetNumbers == 81;
    }

    public int? GetValue(int rowIndex, int columnIndex)
    {
        if (!IsValidIndex(rowIndex) || !IsValidIndex(columnIndex))
        {
            throw new Exception($"Tried accessing invalid position (r:{rowIndex},c:{columnIndex})");
        }

        return _numbers[rowIndex, columnIndex]?.Value;
    }

    public void SetValue(int value, Position position)
    {
        if (_numbers[position.Row, position.Column] is not null)
        {
            throw new Exception($"Value already exists on position {position}.");
        }

        _numbers[position.Row, position.Column] = new(value, true);
        noOfSetNumbers++;
    }

    public static bool IsValidIndex(int index)
        => index >= 0 && index < 9;

    public override string ToString()
    {
        StringBuilder sb = new();

        for(int column = 0; column < 9; column++)
        {
            for(int row = 0; row < 9; row++)
            {
                var valueStr = GetValue(row, column)?.ToString() ?? " ";
                sb.Append(valueStr);

                if (row == 2 || row == 5)
                {
                    sb.Append('|');
                }
            }

            sb.AppendLine();

            if (column == 2 || column == 5)
            {
                sb.Append("-----------");

                sb.AppendLine();
            }
        }

        return sb.ToString();
    }
}

File: Dart/sudoku_console/lib/models/table.dart
import '../models/position.dart';

class Table {
  List<List<int?>> _numbers = List<List<int?>>.generate(
      9, (ind) => List<int?>.generate(9, (index) => null));

  Table({List<List<int?>>? numbers}) {
    if (numbers != null) {
      _numbers = numbers;
    }
  }

  factory Table.load(List<List<int?>> numbers) {
    return Table(numbers: numbers);
  }

  int _filledCount = 0;

  void setValue({required int value, required Position position}) {
    _validateValue(value);

    _numbers[position.rowIndex][position.columnIndex] = value;
    _filledCount++;
  }

  bool get isFull {
    return _filledCount >= 81;
  }

  int? getValue({
    required int rowIndex,
    required int columnIndex,
  }) {
    _validateIndex(rowIndex);
    _validateIndex(columnIndex);

    return _numbers[rowIndex][columnIndex];
  }

  void _validateValue(int value) {
    if (value < 1 || value > 9) {
      throw Exception("Invalid value: $value");
    }
  }

  void _validateIndex(int index) {
    if (!Position.isValidIndex(index)) {
      throw Exception("Invalid index: $index");
    }
  }
}
